src/overflow.ts

```diff
@@ -76,29 +73,34 @@ class OverflowDirective extends AsyncDirective {
 					}
 				});
 
-				onOverflow({ visible, overflowing });
+				this.onOverflow!({ visible, overflowing });
 			},
-			{ root: slot.parentElement, threshold: 1 },
+			{ root: this.slot.parentElement, threshold: 1 },
 		);
 
-		const elements = Array.from(slot.assignedElements()) as HTMLElement[];
-		for (const c of elements) {
-			this.observer.observe(c);
-		}
-	}
-
-	teardownObserver() {
-		this.observer?.disconnect();
-	}
-
-	disconnected() {
-		this.teardownListener();
-		this.teardownObserver();
-	}
-
-	reconnected() {
-		this.setupListener();
-		this.setupObserver();
+		const observe = () => {
+			if (!this.observer || !this.slot) return;
+			const elements = Array.from(
+				this.slot.assignedElements(),
+			) as HTMLElement[];
+			for (const c of elements) {
+				this.observer.observe(c);
+			}
+		};
+
+		observe();
+
+		this.slot.addEventListener('slotchange', observe);
+
+		this.cleanup = () => {
+			if (this.slot) {
+				this.slot.removeEventListener('slotchange', observe);
+			}
+			if (this.observer) {
+				this.observer.disconnect();
+				this.observer = undefined;
+			}
+		};
```

Comment:
You should keep a reference to the element you added the event on, because this.slot might change before you do the cleanup and it will leak listeners.

```ts
const slot = this.slot;
const observer = ...
this.observer = observer;

this.cleanup = () => {
				slot.removeEventListener('slotchange', observe);
				observer.disconnect();
				observer = undefined;
		};
```